sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -632,13 +667,17 @@ object ScalaReflection extends ScalaReflection {
               "cannot be used as field name\n" + walkedTypePath.mkString("\n"))
           }
 
+          // as a field, value class is represented by its underlying type
+          val trueFieldType =
+            if (isValueClass(fieldType)) getUnderlyingTypeOf(fieldType) else fieldType
```

Comment:
shall we update `dataTypeFor` to handle value class, instead of special handling it here?

Reply:
update `dataTypeFor` is not enough since `trueFieldType` is used in code below as well. We can move this into `getConstructorParameters` but then this special handling is still there, so it will not be cleaner.